runtime/cudaq/qis/remote_state.h

```diff
@@ -0,0 +1,103 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#pragma once
+
+#include "common/SimulationState.h"
+#include "cudaq.h"
+#include "cudaq/utils/cudaq_utils.h"
+
+namespace cudaq {
+/// Implementation of `SimulationState` for remote simulator backends.
+// The state is represented by a quantum kernel.
+// For accessor APIs, we may resolve the state to a state vector by executing
+// the kernel on the remote simulator. For overlap API b/w 2 remote states, we
+// can send both kernels to the remote backend for execution and compute the
+// overlap.
+class RemoteSimulationState : public cudaq::SimulationState {
+protected:
+  std::string kernelName;
+  mutable std::unique_ptr<cudaq::SimulationState> state;
+  mutable std::vector<char> argsBuffer;
```

Comment:
I added a code comment in https://github.com/NVIDIA/cuda-quantum/pull/1803/commits/8ab176faf836e7ff11a77bd548c1c91f836fecb3 to document this.

This mainly because of the lazy evaluation. The base API from `cudaq::SimulationState` is expected to be `const` but we would perform evaluation at those points. 

The `argsBuffer` data needs to be passed as non-const pointer for those `launchKernel` evaluation calls as we;ll.